suite/tests/client-interface/drstatecmp-fuzz-app.c

```diff
@@ -0,0 +1,218 @@
+/* **********************************************************
+ * Copyright (c) 2021 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Fuzzing application to stress-test DR with the drstatecmp library.
+ * Only AArch64 is currently supported.
+ */
+
+#include "configure.h"
+#include "dr_api.h"
+#include "opcode_opnd_pairs.h"
+#include "tools.h"
+#include <assert.h>
+#include <setjmp.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <time.h>
+
+#define NUM_INSTS 10000
+
+#define VERBOSE 0
+
+static byte *generated_code;
+static size_t code_size;
+
+static jmp_buf mark;
+static sigjmp_buf sig_mark;
+
+void
+sig_segv_fpe_handler(int signal)
+{
+    siglongjmp(sig_mark, 1);
+    DR_ASSERT_MSG(false, "Shouldn't be reached");
+}
+
+void
+sigill_handler(int signal, siginfo_t *siginfo, void *uctx)
+{
+    ucontext_t *context = (ucontext_t *)uctx;
+    /* Skip illegal instructions. */
+    context->uc_mcontext.pc += 4;
+}
+
+static void
+print_instr_pc(instr_t *instr, byte *encode_pc)
+{
+#if VERBOSE > 0
+    fprintf(stderr, "%p: ", encode_pc);
+    instr_disassemble(GLOBAL_DCONTEXT, instr, STDERR);
+    fprintf(stderr, "\n");
+#endif
+}
+
+static byte *
+append_instr(instr_t *instr, byte *encode_pc)
+{
+    print_instr_pc(instr, encode_pc);
+    byte *nxt_pc = instr_encode(GLOBAL_DCONTEXT, instr, encode_pc);
```

Comment:
Check the return value to ensure it encoded successfully

Reply:
Checked with assertion two lines down but moved it immediately after to make it clearer.